test(create): Add test to check deprecated `apiVersion`s in resource templates created by `helm create`

Bhargav-InfraCloud opened the pull request:
**What this PR does / why we need it**:
1. Change `HorizontalPodAutoscaler`'s version from `autoscaling/v2beta1` -> `autoscaling/v2` in `helm create`.
2. Add a test `TestHelmCreateChart_CheckDeprecatedWarnings` to check the deprecated versions from being created by `helm create`.
3. A comment is added in `pkg/chartutil/create.go` for future manifest files added to `helm create` to follow the same pattern and update the test.

About `TestHelmCreateChart_CheckDeprecatedWarnings`:
- Checks the manifests created by `helm create` to not have deprecated `apiVersion`s.
- This requires the following `ldflags` to be set per the current Kubernetes version for avoiding false positive results.
    1. `-X helm.sh/helm/v3/pkg/lint/rules.k8sVersionMajor=`\<k8s-major-version>
    2. `-X helm.sh/helm/v3/pkg/lint/rules.k8sVersionMinor=`\<k8s-minor-version>
- In order to avoid the `ldflags` passed for this test from conflicting with other tests, a seperate `go test` is added in `make test-unit`.
- Without the `ldflags`, this test passes with a false positive result. This behavior will help in the former run of `go test` and in `make test-coverage`. Note: This test doesn't contribute to coverage anyway.

Closes #11495 

**Special notes for your reviewer**:
Not sure how backward compatibility is checked when the changes are related to the `apiVersion` of the object. But else, I think this change qualifies as backward compatible. Please check.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

@userbradley @joejulian @hzliangbin @wujunwei Please validate whether these changes achieve the expected result. Thanks!

userbradley commented:
ACK -  Will try tomorrow

joejulian reviewed (approved):
Looks good to me. :+1:

vsroy commented:
@joejulian @Bhargav-InfraCloud 

Good Day.
Looks like Kubernetes v1.24 has `End of Life:2023-07-28`
https://kubernetes.io/releases/#release-v1-24

The `autoscaling/v2beta1` API version of HorizontalPodAutoscaler is no longer served as of  Kubernetes v1.25.
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#horizontalpodautoscaler-v125

Can we merge this PR if it looks good ?

joejulian commented:
@vsroy PRs larger than a small require two approvals.

vsroy commented:
@hickeyma @jdolitsky @marckhouzam
Can we have a review on this please.
Based on this -> https://github.com/helm/helm/pull/11923#issuecomment-1566579532
cc @Bhargav-InfraCloud

Vedrillan commented:
The API has changed between v2beta1 and v2. The  `specs.metrics[].Resource` needs to be updated as well if the manifest is upgraded to v2. In particular `targetAverageUtilization` no longer exists.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#resourcemetricsource-v2beta1-autoscaling
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#resourcemetricsource-v2-autoscaling

Bhargav-InfraCloud committed:
test(create): Test to check deprecated resource templates

Test to check deprecation warnings on resource templates that are
created by `helm create` against the latest Kubernetes version. This
test is run in a separate Makefile command to avoid failing other unit
tests configured to run with a specific Kubernetes version(s).

For resources that are disabled by default (like hpa and ingress),
to avoid skipping them from the helm linter test, they are enabled in
the test.

Fixes #11495

Signed-off-by: Bhargav Ravuri <bhargav.ravuri@infracloud.io>

Bhargav-InfraCloud commented:
Thanks! @Vedrillan @vsroy I've updated the PR. Please review.

gjenkins8 commented:
Since [KEP-3136](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/3136-beta-apis-off-by-default/README.md), Kubernetes no longer exposes beta API by default.

For this reason, `helm create` should never need to include a non-permanent API version in the future. Which might make the changes here now mostly redundant?
